### ISSUE TITLE
Support for XZ-compressed files

### ISSUE THREAD

**Issue by fturco** [gjefferyes/RomVault#44]
Let me suggest adding XZ image compression support to RomVault. This would allow saving some space as opposed to using Zip compression.

As a test I tried compressing 26,330 ADF images with both the zip and xz algorithms:

* Uncompressed total size: 22,659 MiB
* Zip-compressed total size: 13,038 MiB
* Xz-compressed total size: 11,596 MiB

So I was able to save 1,442 MiB of space (-11%). That's not a lot but still a significant amount in my opinion.

In my test I used xz default options.

What do you think?

See also https://github.com/FrodeSolheim/fs-uae/issues/124.

**Comment by LnmVolbo** [comment]
Good-to-fair compression stats, but does that produce a solid archive? The benefit of zip / torrentzip is that the already compressed component files can be moved around in blocks, greatly improving rebuilding speed.  When the archive is solid, the slightest change to it requires a full rebuild which really slows down rebuilding.

Further, the benefit of torrentzip is considerable to collectors that distribute collections via torrent.  It's not so much the compression benefit there, it's the benefit for other peers of the torrent to rebuild their partial set to match in order to join the torrent at a high percentage.

I'm not vetoing the idea, just pointing out some considerations why ROMVault currently uses zip and torrentzip.

**Comment by fturco** [comment]
I don't have a deep knowledge of compression algorithms, but according to [this](http://tukaani.org/xz/format.html):

> Random-access reading: The data can be split into independently compressed blocks. Every .xz file contains an index of the blocks, which makes limited random-access reading possible when the block size is small enough.

Is this what you're looking for?

**Comment by LnmVolbo** [comment]
No, the blocks of compression inside each .zip file are not the "blocks" I am talking about.  When a zip file that contains, say a dozen files, and you want your ROM manager to rebuild another zip file that has, say 10 of those files, instead of decompressing all 12 files, then copying the 10 you want, and then compressing the 10, you can just copy the 10 to the new file directly. 

With solid compression, you can only do it the first way.  With zip and torrentzip you can do it either way, the second way being much faster.

 I am calling each file inside the zip archive a "block".  Each of the files compressed inside a zip archive may have their own set of compressed blocks, but they must be taken as a whole set in order to encompass one particular file.

My understanding is that solid compression treats the set of files in it as a single file, and then compresses it.  When it's time to unpack it, there are not areas of the archive corresponding the component compressed blocks of any particular file.  Often the entire solidly compressed file needs to be accessed to retrieve one of the original files.

With a .zip file that has, say 10 files in it of all around the same size each, you can delete the last half of the file and probably recover the first 4 or 5 files.  This is because the first file is in the first 10% of the archive, the second file is in or around the 2nd 10% of the archive, etc.

In a solidly compressed archive, there is no indication of where the parts of any particular file are.  Damaging it, even if not severely, usually results in all files being unrecoverable.

**Comment by artifexor** [comment]
The official [file format description](http://tukaani.org/xz/xz-file-format.txt) does not mentioning any solid feature in xz file. I believe the main problem there is no xz library for C#. You can call xz executable directly of course but that is not so elegant in my opinion.

**Comment by LnmVolbo** [comment]
OK, if you tar archive before compressing, that's a solid archive.  If you compress each file first, then tar archive, that's more like the plain .zip format.

ROMVault currently supports .zip files but only two kinds of compression, those being "Deflate" and "None".  The .zip file format supports other types, such as Deflate64, BZip2, LZMA, & PPMd, but when ROMVault detects these, it reports them as unsupported formats.

**GordonJ** has stated elsewhere that ROMVault will not be supporting any of the other compression formats.  Most of the ROMVault users are supporting torrents and use the torrentzip feature.  torrentzip only uses "Deflate" compression.  There is no reason to use any other format if the files are going to end up being torrentziped anyway.

There are some ROMVault users that are not supporting torrents, of course.  There is another issue (#33) posted a few months ago about not even zipping the files in order to allow the OS to take care of the compression and de-duplication.

**Comment by claunia** [comment]
@LnmVolbo according to what I've understood reading [RomVaultX](https://github.com/gjefferyes/RomVaultX) source code, @gjefferyes is going to implement an opaque depot.

In a nutshell, store every file separately and compressed (he's using gzip) and then offering a way to export files as per the DAT names them, even torrentzipping on request may be possible. I see in his code he's implementing showing the depot as a drive letter on Windows.

This feature backed by xz or filesystem-level-dedup-and-compress will solve both #33 and #44.

**Comment by LnmVolbo** [comment]
@claunia, that sounds interesting. That would fall into making a ROM manager that would also be able to handle .7z files and headers.  I think that ROMVault's current database structure is holding it back from being able to easily handle .7z compression and headers.  I have downloaded the ROMVaultX sourcecode, but I can't get it to completely compile without build errors. It seems to be lacking some bits that make the virtual drive work.  So I haven't been seeing for myself what it can do.

**Comment by claunia** [comment]
@LnmVolbo I don't think the database structure has anything to do about handling 7zip files. I've got RomVaultX to compile using latest VS2015u3, it just complains about a missing project but the code calling it is commented so.

**Comment by LnmVolbo** [comment]
**@claunia**, your description of ROMVaultX was reminiscent of this: https://github.com/uwedeportivo/romba (a Linux / Mac only production so far).

Since the "Vault" is de-duped, it seems to me that the "build output" could be anything from uncompressed folders of files, torrentzipped files, XZ-compressed files, .7z files, or whatever.

Or perhaps I am over-speculating a bit here.  I am assuming the change to the more elaborate database in ROMVaultX is to accomplish the de-duplication.

**Comment by claunia** [comment]
@LnmVolbo more or less I got the impression that ROMVaultX is going to take the way of romba for storing things (in separate gzipped unique files) but in a wayyyyyyy better way (romba developer does not want to support half of the DAT formats, provide a GUI, be multiplatform).

With that storage backend, output can be anything. Complete or partial torrentzipped DATs, torrent7z, etc, etc.
